packages/test/src/test-worker-deployment-versioning.ts

```diff
@@ -0,0 +1,409 @@
+/**
+ * Tests worker-deployment-based versioning
+ *
+ * @module
+ */
+import assert from 'assert';
+import { randomUUID } from 'crypto';
+import asyncRetry from 'async-retry';
+import { ExecutionContext } from 'ava';
+import { Client } from '@temporalio/client';
+import { toCanonicalString, WorkerDeploymentVersion } from '@temporalio/common';
+import { temporal } from '@temporalio/proto';
+import { Worker } from './helpers';
+import { Context, makeTestFunction } from './helpers-integration';
+import { unblockSignal, versionQuery } from './workflows';
+
+const test = makeTestFunction({ workflowsPath: __filename });
+
+test('Worker deployment based versioning', async (t) => {
+  const taskQueue = 'worker-deployment-based-versioning-' + randomUUID();
+  const deploymentName = 'deployment-' + randomUUID();
+  const { client, nativeConnection } = t.context.env;
+
+  const w1DeploymentVersion = {
+    buildId: '1.0',
+    deploymentName,
+  };
+  const w2DeploymentVersion = {
+    buildId: '2.0',
+    deploymentName,
+  };
+  const w3DeploymentVersion = {
+    buildId: '3.0',
+    deploymentName,
+  };
+
+  const worker1 = await Worker.create({
+    workflowsPath: require.resolve('./deployment-versioning-v1'),
+    taskQueue,
+    workerDeploymentOptions: {
+      useWorkerVersioning: true,
+      version: w1DeploymentVersion,
+      defaultVersioningBehavior: 'PINNED',
+    },
+    connection: nativeConnection,
+  });
+  const worker1Promise = worker1.run();
+  worker1Promise.catch((err) => {
+    t.fail('Worker 1.0 run error: ' + JSON.stringify(err));
+  });
```

Comment:
`JSON.stringify(err)` is very likely to simply print out `{}`. You will get more information by just concatenating `err` directly.